tests/integ/test_packaging.py

```diff
@@ -539,3 +576,230 @@ def plus_one_month(x):
+        )
+
+    assert session.get_packages().keys() == {"matplotlib", "numpy", "pyyaml", "scipy"}
+
+    udf_name = Utils.random_name_for_temp_object(TempObjectType.FUNCTION)
+
+    @udf(name=udf_name)
+    def get_numpy_pandas_version() -> str:
+        import skfuzzy as fuzz
+
+        return fuzz.__version__
+
+    Utils.check_answer(session.sql(f"select {udf_name}()"), [Row("0.4.2")])
+
+
+@pytest.mark.skipif(
+    IS_IN_STORED_PROC,
+    reason="subprocess calls are not possible within a stored procedure",
+)
+def test_add_requirements_unsupported_with_persist_path_negative(
+    session, resources_path, temporary_stage
+):
+    """
+    Assert that failure in loading environment from `persist_path` does not affect the requirements addition process,
+    even if requirements are unsupported.
+    """
+    test_files = TestFiles(resources_path)
+
+    with patch.object(session, "_is_anaconda_terms_acknowledged", lambda: True):
+        with patch(
+            "snowflake.snowpark.session.Session._load_unsupported_packages_from_stage",
+            side_effect=Exception("This function does not work"),
+        ):
+            session.add_requirements(
+                test_files.test_unsupported_requirements_file,
+                persist_path=temporary_stage,
+            )
+
+    assert session.get_packages().keys() == {"matplotlib", "numpy", "pyyaml", "scipy"}
+
+    udf_name = Utils.random_name_for_temp_object(TempObjectType.FUNCTION)
+
+    @udf(name=udf_name)
+    def get_numpy_pandas_version() -> str:
+        import skfuzzy as fuzz
+
+        return fuzz.__version__
+
+    Utils.check_answer(session.sql(f"select {udf_name}()"), [Row("0.4.2")])
+
+
+@pytest.mark.skipif(
+    IS_IN_STORED_PROC or IS_WINDOWS,
+    reason="Subprocess calls are not allowed within stored procedures",
+)
+def test_add_requirements_unsupported_with_persist_path(
+    session, resources_path, temporary_stage
+):
+    """
+    Assert that if a persist_path is mentioned, the zipped packages file and a metadata file are present at this
+    remote stage path. Also, subsequent attempts to add the same requirements file should result in the zip file
+    being directly imported from persist_path (i.e. no pip install, no native package dependency detection, etc).
+    We test this by patching the `_upload_unsupported_packages` function to throw an Exception.
+
+    Finally, assert that adding a new unsupported package results in a new environment signature and zip file (i.e.
+    two environments should be present on the stage).
+    """
+    test_files = TestFiles(resources_path)
+
+    # Prove that patching _upload_unsupported_packages leads to failure
+    with patch.object(session, "_is_anaconda_terms_acknowledged", lambda: True):
+        with patch(
+            "snowflake.snowpark.session.Session._upload_unsupported_packages",
+            side_effect=Exception("This function should not have been called"),
+        ):
```

Comment:
Maybe a better side effect message should be `Intentionally raising exception to test`.

Is the exception here raised only to test that we call _upload_unsupported_packages?

Reply:
Yes